File: platforms/ios/Sources/WhiskerModule/WhiskerModuleEventCenter.swift
// iOS module event routing. The active Host installs a Swift closure rather
// than exposing its renderer or FFI implementation to module authors.

import Foundation

/// Shared dispatcher + observer-hook router. All state is internal
/// to the `WhiskerModule` framework; `Module.sendEvent` and the
/// codegen-emitted registration call are the only public entry
/// points.
public enum WhiskerModuleEventCenter {

    private struct EventKey: Hashable {
        let module: String
        let event: String
    }

    private final class SurfaceSink {
        var dispatch: (String, String, WhiskerValue) -> Void
        var observed: Set<EventKey> = []

        init(dispatch: @escaping (String, String, WhiskerValue) -> Void) {
            self.dispatch = dispatch
        }
    }

    // MARK: - Registry

    /// Locked map of `qualifiedName → Module` instance the shared
    /// observer-hook trampolines consult to find the OnStart /
    /// OnStop closure for an incoming `(module, event)` event.
    private static let lock = NSLock()
    private static var modulesByName: [String: Module] = [:]
    private static var eventSinks: [ObjectIdentifier: SurfaceSink] = [:]

    // MARK: - Module registration

    /// Register `module` with the event center. Idempotent — a
    /// second call with the same qualified name replaces the previous
    /// registration (last-write-wins; useful for hot-reload).
    ///
    /// Authors don't call this directly — the codegen-emitted
    /// `_whiskerRegisterModules_<target>()` calls it after
    /// assigning `module.qualifiedName`.
    public static func register(_ module: Module) {
        guard let qname = module.qualifiedName else {
            #if DEBUG
            print("WhiskerModule: register() called on module without qualifiedName — skipped.")
            #endif
            return
        }
        lock.lock()
        modulesByName[qname] = module
        lock.unlock()

    }

    /// Install or remove the event sink owned by one Host surface.
    /// Keeping these as Swift closures avoids a second, legacy C registration
    /// ABI beside WhiskerView's runtime ABI.
    public static func installEventSink(
        owner: AnyObject,
        _ sink: ((String, String, WhiskerValue) -> Void)?
    ) {
        lock.lock()
        let key = ObjectIdentifier(owner)
        var stopped: [EventKey] = []
        if let sink {
            if let existing = eventSinks[key] {
                existing.dispatch = sink
            } else {
                eventSinks[key] = SurfaceSink(dispatch: sink)
            }
        } else {
            let removed = eventSinks.removeValue(forKey: key)
            stopped = removed?.observed.filter { event in
                !eventSinks.values.contains { $0.observed.contains(event) }
            } ?? []
        }
        lock.unlock()
        for event in stopped {
            fireStop(module: event.module, event: event.event)
        }
    }

    /// Update one surface's first/last Rust-side subscription.
    public static func setObserving(
        owner: AnyObject,
        module: String,
        event: String,
        observing: Bool
    ) {
        let key = EventKey(module: module, event: event)
        lock.lock()
        guard let surface = eventSinks[ObjectIdentifier(owner)] else {
            lock.unlock()
            return
        }
        let changed: Bool
        let transition: Bool
        if observing {
            changed = surface.observed.insert(key).inserted
            transition = changed && eventSinks.values.filter { $0.observed.contains(key) }.count == 1
        } else {
            changed = surface.observed.remove(key) != nil
            transition = changed && !eventSinks.values.contains { $0.observed.contains(key) }
        }
        lock.unlock()
        guard transition else { return }
        if observing {
            fireStart(module: module, event: event)
        } else {
            fireStop(module: module, event: event)
        }
    }

    // MARK: - sendEvent

    /// Encode `payload`, dispatch through the bridge, then release
    /// the heap allocations. Called from `Module.sendEvent`.
    internal static func dispatchSend(
        module: String,
        event: String,
        payload: WhiskerValue
    ) {
        let key = EventKey(module: module, event: event)
        lock.lock()
        let sinks = eventSinks.values
            .filter { $0.observed.contains(key) }
            .map(\.dispatch)
        lock.unlock()
        for sink in sinks {
            sink(module, event, payload)
        }
    }

    // MARK: - Observer hook routing

    /// Look up the Module + event-name pair and fire any matching
    /// `OnStartObserving` closures. Called by the shared C
    /// trampoline below.
    private static func fireStart(module: String, event: String) {
        lock.lock()
        let m = modulesByName[module]
        lock.unlock()
        guard let m else { return }
        for hook in m.definitionLazy.onStartObservingHooks where hook.eventName == event {
            hook.handler()
        }
    }

    /// Counterpart to `fireStart`.
    private static func fireStop(module: String, event: String) {
        lock.lock()
        let m = modulesByName[module]
        lock.unlock()
        guard let m else { return }
        for hook in m.definitionLazy.onStopObservingHooks where hook.eventName == event {
            hook.handler()
        }
    }
}

File: platforms/android/module/src/main/kotlin/rs/whisker/runtime/WhiskerModuleEventCenter.kt
// Android module event routing. The active Host installs a Kotlin sink rather
// than requiring modules to know about a JNI or renderer implementation.

package rs.whisker.runtime

import java.util.concurrent.ConcurrentHashMap
import java.util.IdentityHashMap

/** Process-wide module registry and surface-scoped Host event subscriptions. */
public object WhiskerModuleEventCenter {

    private data class EventKey(val module: String, val event: String)

    private class SurfaceSink(
        var dispatch: (String, String, WhiskerValue) -> Unit,
        val observed: MutableSet<EventKey> = mutableSetOf(),
    )

    /**
     * `qualifiedName → Module` lookup the JNI trampolines consult
     * to find the OnStart / OnStop closures for an incoming
     * `(module, event)` event.
     */
    private val modulesByName = ConcurrentHashMap<String, Module>()
    private val eventSinkLock = Any()
    private val eventSinks = IdentityHashMap<Any, SurfaceSink>()

    /**
     * Register [module] after its qualified name has been assigned.
     * Idempotent — re-registering replaces the previous entry.
     */
    @JvmStatic
    public fun register(module: Module) {
        val qname = module.qualifiedName ?: return
        modulesByName[qname] = module
    }

    /** Install or remove the event consumer owned by one Host surface. */
    @JvmStatic
    public fun installEventSink(
        owner: Any,
        sink: ((String, String, WhiskerValue) -> Unit)?,
    ) {
        val stopped = synchronized(eventSinkLock) {
            if (sink == null) {
                val removed = eventSinks.remove(owner) ?: return@synchronized emptyList()
                removed.observed.filterTo(mutableListOf()) { key ->
                    eventSinks.values.none { key in it.observed }
                }
            } else {
                val existing = eventSinks[owner]
                if (existing == null) {
                    eventSinks[owner] = SurfaceSink(sink)
                } else {
                    existing.dispatch = sink
                }
                emptyList()
            }
        }
        stopped.forEach { key -> fireStop(key.module, key.event) }
    }

    /** Update one surface's first/last Rust-side subscription. */
    @JvmStatic
    public fun setObserving(
        owner: Any,
        moduleName: String,
        eventName: String,
        observing: Boolean,
    ) {
        val key = EventKey(moduleName, eventName)
        val transition = synchronized(eventSinkLock) {
            val surface = eventSinks[owner] ?: return
            if (observing) {
                if (!surface.observed.add(key)) return
                eventSinks.values.count { key in it.observed } == 1
            } else {
                if (!surface.observed.remove(key)) return
                eventSinks.values.none { key in it.observed }
            }
        }
        if (transition) {
            if (observing) fireStart(moduleName, eventName) else fireStop(moduleName, eventName)
        }
    }

    /** Forward an event only to surfaces currently observing its channel. */
    internal fun dispatchSend(
        moduleName: String,
        eventName: String,
        payload: WhiskerValue,
    ) {
        val key = EventKey(moduleName, eventName)
        val sinks = synchronized(eventSinkLock) {
            eventSinks.values.filter { key in it.observed }.map { it.dispatch }
        }
        if (sinks.isEmpty()) return
        val snapshot = payload.snapshotForDispatch()
        sinks.forEach { sink -> sink(moduleName, eventName, snapshot) }
    }

    /**
     * Notify the registered module that its listener count changed 0 → 1.
     * Hosts call this when wiring Rust-side subscriptions.
     */
    @JvmStatic
    internal fun fireStart(moduleName: String, eventName: String) {
        modulesByName[moduleName]?.fireOnStartObserving(eventName)
    }

    /** Counterpart to [fireStart] — fires on 1 → 0 transitions. */
    @JvmStatic
    internal fun fireStop(moduleName: String, eventName: String) {
        modulesByName[moduleName]?.fireOnStopObserving(eventName)
    }

}

/** Own mutable byte/container storage before crossing an asynchronous Host turn. */
private fun WhiskerValue.snapshotForDispatch(): WhiskerValue = when (this) {
    WhiskerValue.Null -> this
    is WhiskerValue.Bool -> this
    is WhiskerValue.Int -> this
    is WhiskerValue.Float -> this
    is WhiskerValue.Str -> this
    is WhiskerValue.Bytes -> WhiskerValue.Bytes(value.copyOf())
    is WhiskerValue.Array -> WhiskerValue.Array(value.map { it.snapshotForDispatch() })
    is WhiskerValue.Map -> WhiskerValue.Map(
        value.entries.associate { (key, item) -> key to item.snapshotForDispatch() },
    )
    is WhiskerValue.Err -> this
}
